src/libexpr/flake/call-flake.nix

```
@@ -12,7 +12,36 @@ overrides:
 
 let
 
-  lockFile = builtins.fromJSON lockFileStr;
+  origLockFile = builtins.fromJSON lockFileStr;
+
+  # Build an optimized lock file by consolidating equivalent inputs.
+  lockFile =
+    let
+      nodeCacheKey = key: node:
```

Not sure I would call this a cache.

```suggestion
      nodeDeduplicationKey = key: node:
```

---

Coming to think of it, instead of the various suggestions, it might be helpful to have a generic function for producing a "deduplication map".
A function that behaves like the following seems feasible, and it would cover a large part of what the new code does, and the division of labor should make it easier to understand.

```nix
/**
   Given a function that returns a string key that sufficiently identifies an attribute value,
   return an attribute set that returns a unique attribute name for all equivalent attribute values.

   Type:
      (a -> String) -> AttrsOf a -> AttrsOf String

   Parameters:
     Consider a call `deduplicationMap f m`, then:

     `f` must be a function that returns a string for all attribute values of `m`.

     `m` must be an attribute set.

   Return value:
     An attribute set with the same attribute names as `m`, but each attribute value is a string that represents an attribute name in the original set, whose value produces the same string according to `f`.

   Laws:

       A deduplicated attrset produces the same result under `f`.

       # for all valid f and m:
       let
         dedupMap = deduplicationMap f m;
         dedupped = mapAttrs (k: v: m.${v}) dedupMap;
       in
          mapAttrs (_k: v: f v) m == mapAttrs (_k: v: f v) dedupped
 */
deduplicationMap = f: m: ...
```

I think it deserves to be in `lib` too.